dlmalloc/src/malloc.c

```diff
@@ -4105,7 +4105,15 @@ static void* sys_alloc(mstate m, size_t nb) {
     msegmentptr ss = (m->top == 0)? 0 : segment_holding(m, (char*)m->top);
     ACQUIRE_MALLOC_GLOBAL_LOCK();
 
+#ifdef __wasilibc_unmodified_upstream
     if (ss == 0) {  /* First time through or recovery */
+#else
+    // In wasi-libc, this branch should never be taken as CALL_MORECORE(0) may
+    // lead to incorrect allocations. See try_init_allocator() for more
+    // information.
+    assert(ss != 0);
```

Comment:
but ss can be 0 here if there was not enough space in the initial heap, can't it?

Reply:
@MoritzS Would you be able to comment on this question?

Reply:
That is true, however we still cannot take the if-branch because using `CALL_MORECORE(0)` will almost always be incorrect (and that's the whole point of why `try_init_allocator()` exists). So, maybe we should make sure that `try_init_allocator()` either allocates at least one page if the initial heap is too small, or maybe even directly abort?

Reply:
what's wrong with assuming the discontiguous case if ss == 0?

Reply:
You are right. I initially assumed that `ss` can only be 0 in cases where the memory allocation failed, but that is not the case. I will change the code so that it falls back to the non-contiguous case if `ss` is 0.